feat(biome-js-analyzer): adds class member analyzer to extract mutated properties in a class

vladimir-ivanov opened the pull request:
## Summary

Adds class member analyser that returns all properties mutations in a class member.
Exposes a single fn:

```rust
pub fn class_member_references(list: &JsClassMemberList) -> ClassMemberReferences;

pub struct ClassMemberReferences {
    pub reads: HashSet<ClassMemberReference>,
    pub writes: HashSet<ClassMemberReference>,
}
```

which returns all reads and writes found in any class member (constructors, property and method members).
Different rules apply for extracting writes and reads from a constructor (consistent with eslint)
The code internally collects the `this` keyword and its aliases if any (e.g. const thisAlias = this;) with the scope/ range next to it.
It is useful to avoid matching false positives of this.something in nested constructs.

dyc3 commented on crates/biome_js_syntax/src/class_member_analyzer.rs:
Wouldn't it make more sense for this to go in `biome_js_analyze`? It's called an analyzer

vladimir-ivanov replied:
can move to biome_js_analyze, want to reuse it in 2 rules is the bottom line and enrich with one more fn.
currently  mutated_properties is used in use_readnly_class_properties to extract all mutations of class properties - e.g. this.count++, this.count = 3 etc
It also handles scope aliases e.g.  let scope = this; and then find usages of scope.count++, which no_unused_private_class_properties does not handle 

So the idea is to reuse most of the logic to return another hashset fn readonly_members() for no_unused_private_class_properties rule

dyc3 replied:
> extract all mutations of class properties

Does the semantic model not already do this?

vladimir-ivanov replied:
sadly no. does not operate on classes at all.

dyc3 commented:
I don't really understand what you're trying to get at. I don't have very much context into the problem you're trying to solve.

My comment was asking whether the whole file should go into js_analyze, because at a glance, this does analysis things.

vladimir-ivanov commented:
@ematipico thanks for the feedback.
this pr was more of a draft, probably should have marked it as one. Left it deliberately incomplete/ broken tests etc.

Does the idea make sense to proceed or not at all please? 

It is the main question here. 
Logic is same as found currently in useReadonlyClassProperties:

1. extract all aliases of this and their scopes in a collection
2. find all writes (or mutations) of class property members and constructor properties 
3. (new in this pr) find all reads of class property and function members 

Step 1 is largely needed to resolve a this.scope to aliases e.g. const self = this; and later self.scope is checked against its valid scope.

when 3. is ready we can reuse in no_unused_private_class_members.



Else I can clean up the pr if there is any value of creating this class analyser?

ematipico commented:
You might want to consider expanding the semantic model at this point, or creating a new one just for classes.  I know I said that the current semantic model is more about scopes and such, but after reviewing the code, I think that's the right approach.

ematipico commented:
> Can get rid of them and maybe use empty structs instead? if that is accpetable?

> e.g. struct ThisAliases {}

Yeah that's a valid use-case, however empty struts are usually created because they eventually implement traits.

However, if you design to create a new semantic model for classes, I don't see a compelling argument for using the previous approach. The model will need to save data because consumers will need to query them. For example if you identify a "write", you might want to save the node where it happens (or the text range). Just food for thought

vladimir-ivanov commented on crates/biome_js_analyze/src/class_member_analyzer.rs:
```diff
@@ -0,0 +1,878 @@
+    StatementList(JsStatementList),
+    TemplateElementList(JsTemplateElementList),
+    VariableDeclaratorList(JsVariableDeclaratorList),
+}
```
gone :-) I got confused (now convinced about it) initially that we must create a white list of all nested nodes that we visit (descendants) but if we take this approach maintenance becomes hell.

vladimir-ivanov commented on crates/biome_js_analyze/src/class_member_analyzer.rs:
```diff
@@ -0,0 +1,878 @@
+
+    fn update_fn_body_and_aliases(
+        parent_aliases: &[ClassPropertyReference],
+        results: &mut Vec<ThisAliasesAndTheirScope>,
```
removed entirely, no need for it

vladimir-ivanov commented on crates/biome_js_analyze/src/class_member_analyzer.rs:
```diff
@@ -0,0 +1,878 @@
+    fn write_properties(&self) -> HashSet<ClassPropertyReference>;
+    // class methods and properties
+    fn read_members(&self) -> HashSet<ClassPropertyReference>;
+}
```
removed the trait entirely, thanks

vladimir-ivanov commented on crates/biome_js_analyze/src/class_member_analyzer.rs:
```diff
@@ -0,0 +1,878 @@
+    fn is_this_or_alias(
+        object: &AnyJsExpression,
+        this_aliases: &[ThisAliasesAndTheirScope],
+    ) -> bool;
```
sorted, thanks

vladimir-ivanov commented:
@ematipico I have addressed most of the issues, added comments etc.

now is ok to be looked at.
Personally think that it is much cleaner now. Fixes all the open bugs around no_unused_class_members and reuses logic/ aligns more closely to other biome code?

I would say maybe we make this mergeable.
Then if we still want to do standalone service  alike Semantic or integrate it within, I do not mind, just need a very clear guideance on what exactly to do.

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_readonly_class_properties.rs:
```diff
@@ -134,16 +125,15 @@ declare_lint_rule! {
         let private_only = !ctx.options().check_all_properties;
-        let mut mutated_class_property_names = collect_mutated_class_property_names(&members);
-        mutated_class_property_names.dedup();
+        let ClassMemberReferences { writes, .. } = class_member_references(&members);
```
The answer to your question is here: https://github.com/biomejs/biome/pull/7015#issuecomment-3129461907

ematipico commented on crates/biome_js_analyze/src/class_member_references.rs:
```diff
@@ -0,0 +1,619 @@
+/// - `collect_local_this_aliases`: Collects local aliases of `this` in a function body.
+/// - `collect_all_nested_this_aliases`: Recursively collects `this` aliases in nested function bodies.
+/// - `is_this_or_alias`: Checks if a given expression is a reference to `this` or any of its aliases.
+struct ThisAliasResolver {}
+
+impl ThisAliasResolver {
+    fn collect_local_this_aliases(body: &JsFunctionBody) -> Vec<ClassMemberReference> {
```
As explained before, this isn't usually a pattern used in Rust; however, if you think you're onto something.

Since this type is used to identify the use of `this` inside JavaScript functions, we could save the function to inspect inside the type. e.g. `ThisAliasResolver::new(any_js_function)`. Then the functions will use `any_js_function` as they see fit.

Now, considering the previous suggestion, I find the function `is_this_or_alias` to be out of place. It doesn't do anything that is related to the "`this` within JavaScript functions", which is what `ThisAliasResolver` is for. 

Maybe `is_this_or_alias` should be just its own function, or it should use "javascript functions" somehow

ematipico commented on crates/biome_js_analyze/src/class_member_references.rs:
```diff
@@ -0,0 +1,619 @@
+
+    Some(ClassMemberReferences {
+        reads: reads.into_iter().collect(),
+        writes: writes.into_iter().collect(),
```
```suggestion
        writes: vec![],
```

ematipico reviewed (changes requested):
The new APIs are a bit more clearer now, however the way they play with each other doesn't feel right. I left some suggestions.

Also, could you please add some tests inside `class_member_references`?

vladimir-ivanov commented:
oh, really sorry, will address all the pr comments above then and make it a non draft pr.
Then will raise a ticket to move it to a dedicated service. And ThisScopes can be cleaned up too as part of that ticket.

vladimir-ivanov commented:
> The new APIs are a bit more clearer now, however the way they play with each other doesn't feel right. I left some suggestions.
> 
> Also, could you please add some tests inside `class_member_references`?

done, please let me know if we need more exhaustive tests here.

dyc3 commented on crates/biome_js_analyze/src/class_member_references.rs:
```diff
@@ -0,0 +1,619 @@
+                .and_then(|body| collect_references_from_body(getter.syntax(), &body)),
+            AnyJsClassMember::JsPropertyClassMember(property) => {
+                if let Ok(expression) = property.value()?.expression() {
+                    if let Some(arrow_function) =
+                        JsArrowFunctionExpression::cast(expression.clone().into_syntax())
+                    {
```
also, this comment regarding the clone and cast has not been addressed

dyc3 reviewed (changes requested):
My comment about turning this into a Service was not addressed at all. https://github.com/biomejs/biome/pull/7015#discussion_r2252438248

vladimir-ivanov commented:
> My comment about turning this into a Service was not addressed at all. [#7015 (comment)](https://github.com/biomejs/biome/pull/7015#discussion_r2252438248)

yes, my understanding is that we are doing this as a separate step? Was planning to work on it in 2-3 weeks. Will raise a separate ticket for it if this gets merged. Else will keep this pr and continue then. Thanks

ematipico commented:
Is this a real blocker @dyc3 ? I acknowledged that this should be a service too, but it's just an implementation detail that can be done later, as agreed in the discussion. What do you think?

dyc3 commented:
Sure, a separate PR sounds good to me. I just wanted to make sure that wasn't missed.

vladimir-ivanov commented:
> @vladimir-ivanov
> 
> > also feeling lost on how to address the failure of this test -> pull_diagnostics_of_syntax_rules, think messages are coming in non deterministic order. Changed the test, but not sure if it is the right thing to do. Please advise.
> 
> To be honest, I'm feeling a bit lost about all the changes. The PR description says that it should address only one lint rule, however it seems the scope of the PR has changed. I also thought you were trying to address a couple of bugs of a nursery rule.
> 
> My advice is to take a step back, and keep the changes scoped to that single rule and do not span into others. I really don't understand why the scope changed, and maybe it shouldn't.


@ematipico thanks for the comment.

Reduced the scope of the pr is now ONLY class member analyser. 
Undid the no_unused_private_properties entirely from it.

let me know if we want to keep or scrap entirely the pr please.

Also do we want to move this into service bag at all as discussed above?

if so, do we add it to semantic model or we have separate semantic class model service?

thanks

ematipico commented:
> let me know if we want to keep or scrap entirely the pr please.

There's no need, as long as the scope of the pr doesn't change ☺️ and stays the one you agreed. In fact, you might have noticed that using the new utility in another rule caused a regression, which means the new utility has some bug. 

> Also do we want to move this into service bag at all as discussed above?

Yes I think it should. Do you think it shouldn't? 

> if so, do we add it to semantic model or we have separate semantic class model service?

I think it will be another service, because this one works only for classes
